File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/Cluster-Mode/Python/vserver_tunnel.py
#============================================================
#
# $ID$
#
# vserver_tunnel.py
# Sample code for vserver tunneling
# The API commands are executed on the specified Vserver
# through the Cluster interface.
#
# Copyright 2011 Network Appliance, Inc. All rights
# reserved. Specifications subject to change without notice.
#
#
#============================================================

import sys
sys.path.append("../../../../../lib/python/NetApp")
from NaServer import * 


def print_usage():
    print ("Usage: \nvserver_tunnel [-s] <vserver-name> <storage_system> <user> \n \
                   <password> <ONTAPI-name> [key value] ...\n \
                   -s : Use SSL\n")
    sys.exit (1)


args = len(sys.argv) - 1
dossl = 0
# check for valid number of parameters
#
if(args < 4) :
    print_usage()

opt = sys.argv[1]

if(re.match(r'-', opt)):
    option = opt.split('-')
    storage = sys.argv[3]
    user = sys.argv[4]
    password = sys.argv[5]
    arguments = sys.argv[6:]

    if(option[1] == "s" and args > 2):
        dossl = 1
        vserver = sys.argv[2]

    else:
        print_usage()

else:
    vserver  = opt
    storage = sys.argv[2]
    user = sys.argv[3]
    password = sys.argv[4] 
    arguments = sys.argv[5:]

# open server
server = NaServer(storage, 1, 15)
if(not server.set_vserver(vserver)):
    sys.exit (1)

server.set_admin_user(user, password)

if (dossl) :
    resp = server.set_transport_type("HTTPS")
    if (resp and resp.results_errno() != 0) :
        r = resp.results_reason()
        print ("Unable to set HTTPS transport " + r + "\n")
        sys.exit (1)
length = len(arguments)
if(length > 0):
    # invoke the api with api name and any supplied key-value pairs
    x = NaElement(arguments[0])
    k = 0
    arguments.remove(arguments[0])
    length = length - 2 
    if((length & 1) != 0):
	while(k <= length):
            key = arguments[k]
            k = k + 1
            value = arguments[k] 
            k = k + 1
            x.child_add(NaElement(key,value))
    else:
    	print("Invalid number of parameters")
    	print_usage()
    xo = server.invoke_elem(x)	
else:
    print_usage() 
if ( xo == None ) :
    print ("Invoking api failed to storage_system as user:password.\n")
    sys.exit (1)

# format the output
print ("Output: \n" + xo.sprintf() + "\n" )

File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/Cluster-Mode/Java/vserver_tunnel/vserver_tunnel.java
/*
 * $Id:$
 *
 * vserver_tunnel.java
 * Copyright (c) 2011 NetApp, Inc.
 * All rights reserved.
 *
 * Sample code for vserver tunneling.
 * Any given API is executed on the specified Vserver
 * through the Cluster interface.
 */

import netapp.manage.*;

public class vserver_tunnel {

    public static void usage() {
        System.out
                .println("\nUsage:\n vserver-tunnel [-s] <vserver-name> <storage-system> <user> "
                        + "<password> <ONTAPI-name> [key value] ...");
        System.out.println(" -s : Use SSL\n");
        System.exit(1);
    }

    public static void main(String[] args) {
        NaElement xi;
        NaElement xo;
        NaServer s;
        int index;
        String options;
        int dossl = 0;

        if (args.length < 5) {
            usage();
        }

        index = args[0].indexOf('-');
        if (index == 0 && args[0].charAt(index) == '-') {
            options = args[0].substring(index + 1);
            if (options.equals("s") && args.length > 5) {
                dossl = 1;
            } else {
                usage();
            }
        }

        try {
            if (dossl == 1) {

                // Initialize connection to server, and
                // request version 1.15 of the API set for vserver-tunneling
                //
                s = new NaServer(args[2], 1, 15);
                s.setAdminUser(args[3], args[4]);
                s.setTransportType(NaServer.TRANSPORT_TYPE_HTTPS);
                s.setPort(443);
                s.setVserver(args[1]);

                // Invoke any ONTAPI API with arguments
                // in (key,value) pair
                // args[0]=option,args[1]=vserver-name,args[2]=vserver,
                // args[3]=user,args[4] = passwd, args[5]=Ontapi API,
                // args[6] onwards arguments in (key,value)
                // pair
                //
                xi = new NaElement(args[5]);

                try {
                    if (args.length > 6) {
                        for (int index1 = 6; index1 < args.length; index1++) {
                            xi.addNewChild(args[index1], args[index1 + 1]);
                            index1++;
                        }
                    }
                } catch (ArrayIndexOutOfBoundsException e) {
                    throw new NaAPIFailedException(-1,
                            "Mismatch in arguments passed "
                                    + "(in (key,value) Pair) to "
                                    + "Ontapi API");
                }
            } else {
                // Initialize connection to server, and
                // request version 1.15 of the API set for vserver-tunneling
                //
                s = new NaServer(args[1], 1, 15);
                s.setAdminUser(args[2], args[3]);
                s.setVserver(args[0]);

                // Invoke any ONTAPI API with arguments
                // in (key,value) pair
                // args[0]=filer,args[1]=user,args[2]=passwd
                // args[3]=Ontapi API,args[4] onward arguments
                // in (key,value) pair
                //
                xi = new NaElement(args[4]);
                try {
                    if (args.length > 5) {
                        for (int index2 = 5; index2 < args.length; index2++) {
                            xi.addNewChild(args[index2], args[index2 + 1]);
                            index2++;
                        }
                    }
                } catch (ArrayIndexOutOfBoundsException e) {
                    throw new NaAPIFailedException(-1,
                            "Mismatch in arguments passed "
                                    + "(in (key,value) Pair) to "
                                    + "Ontapi API");
                }
            }
            xo = s.invokeElem(xi);
            System.out.println(xo.toPrettyString(""));
            s.close();
        } catch (NaAPIFailedException e) {
            System.err.println(e.toString());
            System.exit(1);
        } catch (Exception e) {
            System.err.println(e.toString());
            System.exit(1);
        }
    }
}
